src/layers/math/matmul.cpp

```diff
@@ -164,32 +192,35 @@ void fp_compute_impl(matmul_layer<TensorDataType, data_layout::DATA_PARALLEL,El:
   const auto input0_dims = l.get_input_dims(0);
   const auto input1_dims = l.get_input_dims(1);
   const auto output_dims = l.get_output_dims();
+  // Check if matrix is 3D or 2D 
+  const El::Int mat_depth =   (input0_dims.size()>2) ? *(input0_dims.rbegin()+2) : 1;
+
   const El::Int input0_height = *(input0_dims.rbegin()+1);
   const El::Int input0_width = *(input0_dims.rbegin());
   const El::Int input1_width = *(input1_dims.rbegin());
   const El::Int output_height = *(output_dims.rbegin()+1);
   const El::Int output_width = *(output_dims.rbegin());
 
-  // Compute matrix multiplication for each mini-batch sample
-  // Note: cuBLAS expects matrices in Fortran layout while LBANN
-  // tensors are in C layout.
+  const auto num_matrices = mat_depth * local_mini_batch_size;
+  
   {
     using namespace hydrogen;
     auto multisync = MakeMultiSync(gpu::get_sync_info(local_output),
                                    gpu::get_sync_info(local_input0),
                                    gpu::get_sync_info(local_input1));
+
     gpu_blas::GemmStridedBatched(
       transpose_input1 ? TransposeMode::TRANSPOSE : TransposeMode::NORMAL,
       transpose_input0 ? TransposeMode::TRANSPOSE : TransposeMode::NORMAL,
       output_width,
       output_height,
       transpose_input0 ? input0_height : input0_width,
       El::TypeTraits<TensorDataType>::One(),
-      local_input1.LockedBuffer(), input1_width, local_input1.LDim(),
-      local_input0.LockedBuffer(), input0_width, local_input0.LDim(),
+      local_input1.LockedBuffer(), input1_width, local_input1.LDim() / mat_depth,
+      local_input0.LockedBuffer(), input0_width, local_input0.LDim() / mat_depth,
       El::TypeTraits<TensorDataType>::Zero(),
-      local_output.Buffer(), output_width, local_output.LDim(),
-      local_mini_batch_size,
+      local_output.Buffer(), output_width, local_output.LDim() / mat_depth,
```

Comment:
This won't be correct if the matrices aren't contiguous. I'd recommend replacing the matrix strides from `LDim()/mat_depth` to `width*height`. In case we get non-contiguous data, we can have something like:
```c++
if (!local_input1.Contiguous() || !local_input2.Contiguous() || !local_output.Contiguous()) {
  LBANN_ERROR(
    this->get_type()," layer \"",this->get_name(),"\" ",
    "has non-contiguous data buffers");
}
```
We can implement this case properly if we run into it.

Reply:
I'll change it to `width*height`, that should've been written that way. Actually, while we have contiguous buffers, that would make it more general and have MatMul defined for two matrices with shapes ( *, m, k) and  ( *, k, n), and mat1.shape[:-2] == mat2.shape[:-2]. It'd only change one value in the GEMM kernel, I believe.  CPU implementation would need a few changes though.

Reply:
I'm thinking of situations where the Hydrogen matrices have `LDim() != Height()`, so the columns are not contiguous in memory. I believe the CPU implementation handles this case correctly, it's just the GPU implementation. If we wanted to support it, I believe cuBLAS exposes an API for batched GEMM with non-uniform strides, but that would be more effort than it would be worth right now.